src/libostree/ostree-impl-system-generator.c

```diff
@@ -97,14 +97,33 @@ stateroot_from_ostree_cmdline (const char *ostree_cmdline, GError **error)
   static gsize regex_initialized;
   if (g_once_init_enter (&regex_initialized))
     {
-      regex = g_regex_new ("^/ostree/boot.[01]/([^/]+)/", 0, 0, NULL);
+      regex = g_regex_new ("/ostree/[a-z01.]+/([^/]+)/", 0, 0, NULL);
       g_assert (regex);
       g_once_init_leave (&regex_initialized, 1);
     }
 
+  /* The to_parse string was added to handle the aboot case, in the aboot
+   * case it has to parse something like:
```

Comment:
In latest git main, `ostree-prepare-root.c` links to glib at least, so if it helps us we can probably factor out a shared helper more easily between these two.

Reply:
If there's a way we can keep the aboot case more consistent here by setting things up in prepare root that seems nicer to me.

(Can we have prepare-root create a symlink in `/run` with the data it parsed?)

Reply:
The way it is, prepare root just goes to a previously set up symlink at deploy time, which would look something like this:

```
[root@localhost ~]# ls -tlr /ostree/
total 20
lrwxrwxrwx. 1 root root   87 Jun 29 12:34 root.b -> deploy/centos/deploy/6ce6fe7c07c4274f7d536f6d505b34fc1453978f576c4b35d241ab7a4dd06898.0
drwxr-xr-x. 3 root root 4096 Jun 29 12:36 boot.1.1
lrwxrwxrwx. 1 root root    8 Jun 29 12:36 boot.1 -> boot.1.1
lrwxrwxrwx. 1 root root   87 Jun 29 12:37 root.a -> deploy/centos/deploy/76fee8c28c91f92d598217254fabf3da17a309b4208bdc38a135ac54fc1e82de.0
```

was thinking of changing the symlink to ../ostree/deploy/centos/deploy/76fee8c28c91f92d598217254fabf3da17a309b4208bdc38a135ac54fc1e82de.0 to help satisfy the regex.

But I wasn't taking the generator into account at the time. /run likely wouldn't work as it wouldn't persist across boots?

I'm open to ideas for an alternate location if another location makes more sense...

Reply:
> If there's a way we can keep the aboot case more consistent here by setting things up in prepare root that seems nicer to me.
> 
> (Can we have prepare-root create a symlink in `/run` with the data it parsed?)

So initially root.a and root.b would be symlinks with the exact same values as the karg for ostree= in the grub case, so it was the exact code flow once you resolved the symlink (with boot.1 or boot.0, etc.). But then I saw that sometimes that lead to broken links in rollback cases. So now it's almost the same except now you have to resolve the symlink and tweak the regex just a tiny bit.

So now we use the realpaths instead for the root.a and root.b, the end result is the same once, you parse it a little differently here. I've tested with your review comments addressed and it's all working perfectly now (the selinux issue was fixed somewhere).

I would like to try other approaches, but I think other approaches would lead to more complex changes than seen here. LoC is a poor metric, but with these 14 lines of code everything works as expected from my testing and the grub case is not effected.

Reply:
> In latest git main, `ostree-prepare-root.c` links to glib at least, so if it helps us we can probably factor out a shared helper more easily between these two.

So the only role for this function seems to be to get the osname from the path for this line:

https://github.com/ostreedev/ostree/blob/6056ec130ffff4ac310583f4f4ebffef64919fc3/src/libostree/ostree-impl-system-generator.c#L206

does prepare-root need this functionality? I do understand the desire for a generic parser though, you could have a struct maybe that contains these fields.

```
centos 
76b0919b393aa2254dd4b72950514422bb213992c3885f2196ec9cb278c57c5a
0
```

and once you have those fields copied in a struct, just do everything generically based on those fields. It's a bigger rewrite than here though.